owa/app/js/components/manage-autogeneration-option/ManageAutogenerationOption.jsx

```diff
@@ -0,0 +1,455 @@
+    const selectedGenOption = genOptions[id];
+    selectedGenOption.idType = identifierType;
+    selectedGenOption.location = currentLocation;
+    selectedGenOption.sourceName = autogenerateSource;
+    selectedGenOption.autoGen = autoGenerationEnabled;
+    selectedGenOption.manualEntry = manualEntryEnabled;
+    genOptions[id] = selectedGenOption; 
+    this.setState({
+      genOptions,
+      modal: false
+    });
+  }
+
+  onChange(e) {
+    let currentGenOption;
+    if (e.target.name === 'autoGenerationEnabled' || e.target.name === 'manualEntryEnabled') {
+      currentGenOption = Object.assign({}, this.state.currentGenOption, { [e.target.name]: e.target.checked});
+    } else {
+      currentGenOption = Object.assign({}, this.state.currentGenOption, { [e.target.name]: e.target.value});
+    }
+    this.setState({
+      alertVisible: false,
+      currentGenOption
+    })
+  }
+
+  deleteGenOption(id) {
+    console.log('delete', id);
+  }
+
+  createGenOptions() {
+    this.setState({
+      saveGenOptionStatus: 'new',
+      modal: true,
+      firstLevelCompleted: false,
+      currentGenOption: {
+        identifierType: '',
+        currentLocation: '',
+        autogenerateSource: '',
+        autoGenerationEnabled: false,
+        manualEntryEnabled: false
+      }
+    });
+  }
+
+  toggle() {
+    this.setState({
+      modal: !this.state.modal
+    });
+  }
+
+  editGenOption(id) {
+
+  }
+
+  nextStage() {
+    const currentGenOption = this.state.currentGenOption;
+    if (currentGenOption.identifierType === '') {
+      this.setState({
+        errors: {
+          message: 'Please select an identifier type'
+        },
+        alertVisible: true,
+      })
+    } else {
+      this.setState({
+        firstLevelCompleted: true,
+        currentGenOption: {
+          identifierType: currentGenOption.identifierType,
+          currentLocation: currentGenOption.currentLocation,
+          autogenerateSource: currentGenOption.autogenerateSource,
+          autoGenerationEnabled: currentGenOption.autoGenerationEnabled,
+          manualEntryEnabled: currentGenOption.manualEntryEnabled
+        }
+      });
+    }
+    
+  }
+
+  onDismiss() {
+    this.setState({ alertVisible: false });
+  }
+
+  saveNewAutoGenOptions() {
+    const { currentGenOption, genOptions } = this.state;
+    if (currentGenOption.currentLocation === '' || currentGenOption.autogenerateSource === '') {
+      this.setState({
+        errors: {
+          message: 'All fields are required',
+        },
+        alertVisible: true
+      })
+    } else {
+      const newIdentifierType = {
+        idType: currentGenOption.identifierType,
+        location: currentGenOption.currentLocation,
+        sourceName: currentGenOption.autogenerateSource,
+        manualEntry: currentGenOption.manualEntryEnabled,
+        autoGen: currentGenOption.autoGenerationEnabled
+      }
+      genOptions.push(newIdentifierType);
+      this.setState({
```

Comment:
Read [this](https://medium.com/@shopsifter/using-a-function-in-setstate-instead-of-an-object-1f5cfd6e55d1) it explains why it's best not to set the state with an object.